File: 2021/day02/day021.hs
module Main where

main :: IO ()
main = interact
  $ show
  . uncurry (*)
  . foldl solve (0, 0)
  . lines

solve :: (Int, Int) -> String -> (Int, Int)
solve (hor, dpt) l = 
  case dir of
    "forward" -> (hor+n, dpt  )
    "down"    -> (hor  , dpt+n)
    "up"      -> (hor  , dpt-n)
  where 
    [dir, sn] = words l
    n = read sn

File: 2021/day02/initial_solution/day021.py
import sys
from functools import reduce 

def solve(sub, l):
  hor, dpt, aim = sub
  dir, sn = l
  n = int(sn)
  return (
    (dir == "forward")*[hor + n, dpt + n*aim, aim  ] + 
    (dir == "down")   *[hor    , dpt        , aim+n] + 
    (dir == "up")     *[hor    , dpt        , aim-n] )

inp = [line.split() for line in sys.stdin.readlines()]
hor, dpt, _ = reduce(solve, inp, [0,0,0])
print(hor * dpt)